File: src/TradingSystem.Web/Services/CurrencyService.cs
using System.Net.Http.Json;
using Microsoft.JSInterop;
using TradingSystem.Web.Models;

namespace TradingSystem.Web.Services;

public interface ICurrencyService
{
    string CurrentCurrencyCode { get; }
    string CurrentCurrencySymbol { get; }
    decimal CurrentRate { get; }
    List<CurrencyInfo> AvailableCurrencies { get; }
    event Action? OnCurrencyChanged;
    Task LoadCurrenciesAsync();
    Task SetCurrencyAsync(string currencyCode);
    string FormatPrice(decimal usdValue);
    /// <summary>
    /// Format a price that is in the given source currency, converting to the user's selected display currency.
    /// </summary>
    string FormatPrice(decimal value, string sourceCurrency);
    decimal ConvertFromUsd(decimal usdValue);
    /// <summary>
    /// Convert a value from the given source currency to the user's selected display currency.
    /// </summary>
    decimal Convert(decimal value, string sourceCurrency);
}

public class CurrencyService : ICurrencyService
{
    private readonly HttpClient _httpClient;
    private readonly IJSRuntime _jsRuntime;
    private const string CurrencyKey = "selectedCurrency";

    public string CurrentCurrencyCode { get; private set; } = "USD";
    public string CurrentCurrencySymbol { get; private set; } = "$";
    public decimal CurrentRate { get; private set; } = 1.0m;
    public List<CurrencyInfo> AvailableCurrencies { get; private set; } = new();

    public event Action? OnCurrencyChanged;

    public CurrencyService(HttpClient httpClient, IJSRuntime jsRuntime)
    {
        _httpClient = httpClient;
        _jsRuntime = jsRuntime;
    }

    public async Task LoadCurrenciesAsync()
    {
        try
        {
            var response = await _httpClient.GetFromJsonAsync<ApiResponse<CurrencyListResponse>>("/currency");
            if (response?.Success == true && response.Data != null)
            {
                AvailableCurrencies = response.Data.Currencies;
            }
        }
        catch
        {
            // Fallback currencies
            AvailableCurrencies = new List<CurrencyInfo>
            {
                new() { Code = "USD", Symbol = "$", Name = "US Dollar", RateFromUsd = 1.0m },
                new() { Code = "INR", Symbol = "₹", Name = "Indian Rupee", RateFromUsd = 83.50m },
                new() { Code = "EUR", Symbol = "€", Name = "Euro", RateFromUsd = 0.92m },
                new() { Code = "GBP", Symbol = "£", Name = "British Pound", RateFromUsd = 0.79m }
            };
        }

        // Load saved preference
        try
        {
            var savedCurrency = await _jsRuntime.InvokeAsync<string?>("localStorage.getItem", CurrencyKey);
            if (!string.IsNullOrEmpty(savedCurrency))
            {
                var currency = AvailableCurrencies.FirstOrDefault(c => c.Code == savedCurrency);
                if (currency != null)
                {
                    CurrentCurrencyCode = currency.Code;
                    CurrentCurrencySymbol = currency.Symbol;
                    CurrentRate = currency.RateFromUsd;
                }
            }
        }
        catch
        {
            // Default to USD
        }
    }

    public async Task SetCurrencyAsync(string currencyCode)
    {
        var currency = AvailableCurrencies.FirstOrDefault(c => c.Code == currencyCode);
        if (currency == null) return;

        CurrentCurrencyCode = currency.Code;
        CurrentCurrencySymbol = currency.Symbol;
        CurrentRate = currency.RateFromUsd;

        try
        {
            await _jsRuntime.InvokeVoidAsync("localStorage.setItem", CurrencyKey, currencyCode);
        }
        catch { }

        OnCurrencyChanged?.Invoke();
    }

    public string FormatPrice(decimal usdValue)
    {
        var converted = usdValue * CurrentRate;
        return $"{CurrentCurrencySymbol}{converted:N2}";
    }

    public string FormatPrice(decimal value, string sourceCurrency)
    {
        var converted = Convert(value, sourceCurrency);
        return $"{CurrentCurrencySymbol}{converted:N2}";
    }

    public decimal ConvertFromUsd(decimal usdValue)
    {
        return usdValue * CurrentRate;
    }

    public decimal Convert(decimal value, string sourceCurrency)
    {
        if (string.IsNullOrEmpty(sourceCurrency))
            sourceCurrency = "USD";

        // If source currency == display currency, no conversion needed
        if (sourceCurrency.Equals(CurrentCurrencyCode, StringComparison.OrdinalIgnoreCase))
            return value;

        // Step 1: Convert source → USD (divide by source rate)
        var sourceRate = GetRateForCurrency(sourceCurrency);
        var valueInUsd = value / sourceRate;

        // Step 2: Convert USD → target (multiply by target rate)
        return valueInUsd * CurrentRate;
    }

    private decimal GetRateForCurrency(string currencyCode)
    {
        var currency = AvailableCurrencies.FirstOrDefault(c =>
            c.Code.Equals(currencyCode, StringComparison.OrdinalIgnoreCase));
        return currency?.RateFromUsd ?? 1.0m;
    }
}

File: src/TradingSystem.Angular/src/app/services/currency.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { BehaviorSubject, Observable, of } from 'rxjs';
import { map, catchError, tap } from 'rxjs/operators';
import { ApiResponse, CurrencyInfo, CurrencyListResponse } from '../models/models';

@Injectable({ providedIn: 'root' })
export class CurrencyService {
  private readonly STORAGE_KEY = 'selectedCurrency';
  private currenciesSubject = new BehaviorSubject<CurrencyInfo[]>([]);
  private currentCurrency: CurrencyInfo = { code: 'USD', symbol: '$', name: 'US Dollar', rateFromUsd: 1 };

  currencies$ = this.currenciesSubject.asObservable();

  get currentCode(): string { return this.currentCurrency.code; }
  get currentSymbol(): string { return this.currentCurrency.symbol; }
  get currentRate(): number { return this.currentCurrency.rateFromUsd; }
  get availableCurrencies(): CurrencyInfo[] { return this.currenciesSubject.value; }

  constructor(private http: HttpClient) {}

  loadCurrencies(): Observable<CurrencyInfo[]> {
    return this.http.get<ApiResponse<CurrencyListResponse>>('/currency').pipe(
      map(r => r.data?.currencies || this.getDefaultCurrencies()),
      tap(currencies => {
        this.currenciesSubject.next(currencies);
        const saved = localStorage.getItem(this.STORAGE_KEY);
        if (saved) {
          const found = currencies.find((c: CurrencyInfo) => c.code === saved);
          if (found) this.currentCurrency = found;
        }
      }),
      catchError(() => {
        const defaults = this.getDefaultCurrencies();
        this.currenciesSubject.next(defaults);
        return of(defaults);
      })
    );
  }

  setCurrency(code: string): void {
    const found = this.currenciesSubject.value.find(c => c.code === code);
    if (found) {
      this.currentCurrency = found;
      localStorage.setItem(this.STORAGE_KEY, code);
    }
  }

  formatPrice(usdValue: number, sourceCurrency: string = 'USD'): string {
    const converted = this.convert(usdValue, sourceCurrency);
    return `${this.currentSymbol}${converted.toFixed(2)}`;
  }

  convert(value: number, sourceCurrency: string = 'USD'): number {
    if (sourceCurrency === this.currentCode) return value;
    // Convert to USD first, then to target
    const sourceInfo = this.currenciesSubject.value.find(c => c.code === sourceCurrency);
    const usdValue = sourceInfo ? value / sourceInfo.rateFromUsd : value;
    return usdValue * this.currentRate;
  }

  private getDefaultCurrencies(): CurrencyInfo[] {
    return [
      { code: 'USD', symbol: '$', name: 'US Dollar', rateFromUsd: 1 },
      { code: 'INR', symbol: '₹', name: 'Indian Rupee', rateFromUsd: 83.5 },
      { code: 'EUR', symbol: '€', name: 'Euro', rateFromUsd: 0.92 },
      { code: 'GBP', symbol: '£', name: 'British Pound', rateFromUsd: 0.79 }
    ];
  }
}
